### PR TITLE
135 Model Tests

### LINKED ISSUE #135: Increase Test Coverage on Models

### PR DESCRIPTION
**Pull request by JPrevost**
Along with the open dashboard PR (which has the Subscriber tests), this provides around 95% coverage for our Models with some exceptions:

- HubUtils (a pseudomodel, so I'll tackle separately once we confirm if that's the best place for those kinds of features)
- User (which is about to get serious in the coming weeks so testing it as is seems pointless)
- Validator (I don't think we really use this yet so I was reluctant to build tests for it)

It's possible a few tests are flakey as they do things based on time. If they do indeed become flakey (passing sometimes, failing others) we can temporarily skip them, figure out how to mock time, or allow some of our create methods to accept a Date of creation (but still default to Now if none is passed). I think they should be okay as I put in `Thread.sleep`s where needed...which is horrible, but works (maybe).

I'll mark some specific changes for closer inspection.

closes #135

### CODE REVIEW

File: app/models/ResourceMap.scala [hunk 1]
```diff
@@ -37,7 +37,7 @@ case class ResourceMap(id: Int, tag: String, description: String, swordUrl: Opti
 
   def schemes: List[Scheme] = {
     DB.withConnection { implicit c =>
-      SQL("select scheme.* from scheme, resource_map_scheme, resource_map where scheme.id = resource_map_scheme.scheme_id and resource_map_scheme.resource_map_id = resource_map.id and resource_map.id = {resmap_id}")
+      SQL("select distinct scheme.id, scheme.* from scheme, resource_map_scheme, resource_map where scheme.id = resource_map_scheme.scheme_id and resource_map_scheme.resource_map_id = resource_map.id and resource_map.id = {resmap_id}")
```
Comment (JPrevost) [inline app/models/ResourceMap.scala]: I opened ticket #139 and made the required change here. If this is incorrect, let me know but it made sense to me this way.

File: app/models/Subscription.scala [hunk 2]
```diff
@@ -88,6 +88,8 @@ object Subscription {
     }
   }
 
+  // todo: maybe a better name would be schemeTopicCount? The current name makes me expect a count
+  // of schemes... which would always be 1 as we pass in the Scheme to count.
   def schemeCount(subscriberId: Int, schemeId: Int) = {
```
Comment (JPrevost) [inline app/models/Subscription.scala]: As my comment says in the code, this method name is confusing to me.

File: app/models/Subscription.scala [hunk 3]
```diff
@@ -111,9 +113,9 @@ object Subscription {
   }
 
   def create(subscriberId: Int, topicId: Int, action: String, earliest: Date, latest: Date) = {
-    val created = new Date
-    val updated = created
-    val cancelled = created
+    val created = new Date  // todo: should db defaults should handle this?
+    val updated = created   // todo: should db defaults should handle this?
+    val cancelled = created // todo: null might be better?
```
Comment (JPrevost) [inline app/models/Subscription.scala]: I didn't change this, but we do let the DB handle this for other models.

File: app/models/Topic.scala [hunk 4]
```diff
@@ -158,9 +158,9 @@ object Topic {
     }
   }
 
-  def deleteUnlinkedBefore(date: Date) = {
+  def deleteUnlinkedBefore(date: Date) {
     DB.withConnection { implicit c =>
-      SQL("delete from topic where created < {created}").on('created -> date)
+      SQL("delete from topic where created < {created}").on('created -> date).executeUpdate
```
Comment (JPrevost) [inline app/models/Topic.scala]: I enabled this query to execute. It'll still runtime error if called on Topics with linked Items due to foreign key restrictions.

See https://github.com/MITLibraries/scoap3hub/issues/141

File: test/unit/ContentProfileSpec.scala [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        Scheme.create("tag", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        ContentProfile.findById(1).get.schemes.size must equalTo(0)
+
+        ContentProfile.findById(1).get.addScheme(Scheme.findByTag("tag").get)
+        ContentProfile.findById(1).get.schemes.size must equalTo(1)
+
+        ContentProfile.findById(1).get.addScheme(Scheme.findByTag("tag").get)
+        ContentProfile.findById(1).get.schemes.size must equalTo(1)
+      }
+    }.pendingUntilFixed("See https://github.com/MITLibraries/scoap3hub/issues/136")
```
Comment (JPrevost) [inline test/unit/ContentProfileSpec.scala]: See https://github.com/MITLibraries/scoap3hub/issues/136

File: test/unit/ResourceMapSpec.scala [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      }
+    }
+
+    "#create does not require a sword url" in {
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        ResourceMap.all.size must equalTo(0)
+        ResourceMap.create("tag", "desc", None)
+        ResourceMap.all.size must equalTo(1)
+      }
+    }.pendingUntilFixed("See https://github.com/MITLibraries/scoap3hub/issues/137")
```
Comment (JPrevost) [inline test/unit/ResourceMapSpec.scala]: See https://github.com/MITLibraries/scoap3hub/issues/137

File: test/unit/ResourceMapSpec.scala [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        val s = Scheme.make("tag", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val cf = ContentFormat.make("tag", "label", "desc", "http://www.example.com", "mimetype", Some("logo"))
+        val rm = ResourceMap.make("tag", "desc", Some("http://www.example.com"))
+
+        rm.mappingsForScheme(s).size must equalTo(0)
+        rm.addMapping(s.id, cf.id, "source", 1)
+        rm.addMapping(s.id, cf.id, "source", 1) must throwA[Exception].like {
+                case e: Exception => e.getMessage aka "error" mustEqual (
+                  "JdbcSQLException(Unique index or primary key violation)")}
+      }.pendingUntilFixed("https://github.com/MITLibraries/scoap3hub/issues/138")
```
Comment (JPrevost) [inline test/unit/ResourceMapSpec.scala]: https://github.com/MITLibraries/scoap3hub/issues/138

File: test/unit/SubscriptionSpec.scala [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+      running(FakeApplication(additionalConfiguration = inMemoryDatabase())) {
+        val u = User.make("bob", "bob@example.com", "pwd", "role1")
+        val sub = Subscriber.make(u.id, "Sub Name", "cat", "contact", Some("link"), Some("logo"))
+        val scheme = Scheme.make("tag", "gentype", "cat", "desc", Some("link"), Some("logo"))
+        val t = Topic.make(scheme.id, "tag", "name")
+        Subscription.schemeCount(sub.id, scheme.id) must equalTo(0)
+        Subscription.create(sub.id, t.id, "deliver", sub.created, sub.created) must throwA[Exception].like {
+                case e: Exception => e.getMessage aka "error" mustEqual (
+                  "JdbcSQLException(Unique index or primary key violation)")}
+      }.pendingUntilFixed("https://github.com/MITLibraries/scoap3hub/issues/140")
```
Comment (JPrevost) [inline test/unit/SubscriptionSpec.scala]: See https://github.com/MITLibraries/scoap3hub/issues/140

File: test/unit/TopicSpec.scala [hunk 9]
```diff
@@ -83,45 +99,297 @@ class TopicSpec extends Specification {
...
+        val rmap = ResourceMap.make("rmaptag", "rmapdesc", Some(""))
+        val col = Collection.make(pub.id, ct.id, rmap.id, "coll", "desc", "open")
+        val item = Item.make(col.id, ct.id, "location", "scoap:abc:123")
+        val item2 = Item.make(col.id, ct.id, "location", "scoap:abc:456")
+        val item3 = Item.make(col.id, ct.id, "location", "scoap:abc:789")
+        item.addTopic(t1)
+
+        Topic.deleteUnlinkedBefore(new Date)
+        Topic.all must haveSize(1)
+      }.pendingUntilFixed(": See https://github.com/MITLibraries/scoap3hub/issues/141")
```
Comment (JPrevost) [inline test/unit/TopicSpec.scala]: See https://github.com/MITLibraries/scoap3hub/issues/141

File: app/models/ResourceMap.scala (hunk 1 above)
Comment (richardrodgers) [inline app/models/ResourceMap.scala]: Nope, makes sense to me as well

File: app/models/Subscription.scala (hunk 3 above)
Comment (richardrodgers) [inline app/models/Subscription.scala]: Yea, but these might work differently in the future. The date handling around subscriptions will require attention at some point, so the comment is very useful

File: app/models/Topic.scala (hunk 4 above)
Comment (richardrodgers) [inline app/models/Topic.scala]: Seems fine

**Comment by richardrodgers** [comment]
The other issues can be addressed through their respective issues. Great to get so much additional test coverage. :+1: :+1:

File: app/models/Subscription.scala (hunk 3 above)
Comment (JPrevost) [inline app/models/Subscription.scala]: Gotcha, makes sense. Thanks.

**Pull request merged**